kernel/futex.c

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) 2019 Intel corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <spinlock.h>
+#include <kswap.h>
+#include <wait_q.h>
+#include <syscall_handler.h>
+#include <init.h>
+#include <ksched.h>
+
+static struct k_spinlock lock;
+
+static _wait_q_t *k_futex_find_waitq(struct k_futex *futex)
+{
+	struct _k_object *obj;
+
+	obj = z_object_find(futex);
+	if (obj == NULL || obj->type != K_OBJ_FUTEX) {
+		return NULL;
+	}
+
+	return (_wait_q_t *)obj->data;
+}
+
+void z_impl_k_futex_init(struct k_futex *futex)
+{
+	futex->val = 0U;
+	z_object_init(futex);
+}
+
+#ifdef CONFIG_USERSPACE
+Z_SYSCALL_HANDLER(k_futex_init, futex)
+{
+	if (Z_SYSCALL_MEMORY_WRITE(futex, sizeof(struct k_futex)) != 0) {
+		return -EACCES;
+	}
+
+	z_impl_k_futex_init((struct k_futex *)futex);
+
+	return 0;
+}
+#endif
+
+int z_impl_k_futex_wake(struct k_futex *futex, unsigned int num_waiters)
+{
+	k_spinlock_key_t key;
+	_wait_q_t *thread_q;
+	struct k_thread *thread;
+	unsigned int woken = 0;
+
+	thread_q = k_futex_find_waitq(futex);
+	if (thread_q == NULL) {
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&lock);
+
+	while (woken < num_waiters &&
+	    (thread = z_unpend_first_thread(thread_q))) {
+		z_ready_thread(thread);
+		z_set_thread_return_value(thread, 0);
+		woken++;
+	}
+
+	if (woken) {
+		z_reschedule(&lock, key);
```

Comment:
To elaborate here: there **is** a behavior difference between the two strategies: in one, a cooperative caller will only yield if it actually woke up threads on the futex.  In the other, it will **also** yield if anything else in the system happened to become runnable.

I just argue the latter case is too subtle and dangerous to be worth documenting for app authors and that we should just say "k_futex_wait is a reschedule point". And if we do that, we should pick the simpler implementation here.

Reply:
agree, thanks